flink-libraries/flink-table/src/main/scala/org/apache/flink/api/table/typeinfo/NullAwareComparator.scala

```diff
@@ -0,0 +1,231 @@
+      }
+    }
+  }
+
+  override def invertNormalizedKey(): Boolean = wrappedComparator.invertNormalizedKey()
+
+  override def supportsSerializationWithKeyNormalization(): Boolean = false
+
+  override def writeWithKeyNormalization(record: T, target: DataOutputView): Unit =
+    throw new UnsupportedOperationException("Record serialization with leading normalized keys" +
+      " not supported.")
+
+  override def readWithKeyDenormalization(reuse: T, source: DataInputView): T =
+    throw new UnsupportedOperationException("Record deserialization with leading normalized keys" +
+      " not supported.")
+
+  override def isNormalizedKeyPrefixOnly(keyBytes: Int): Boolean =
+    wrappedComparator.isNormalizedKeyPrefixOnly(keyBytes - 1)
+
+  override def setReference(toCompare: T): Unit = {
+    if (toCompare == null) {
+      nullReference = true
+    }
+    else {
+      nullReference = false
+      wrappedComparator.setReference(toCompare)
+    }
+  }
+
+  override def compare(first: T, second: T): Int = {
+    // both values are null -> equality
+    if (first == null && second == null) {
+      0
+    }
+    // first value is null -> inequality
+    // but order is considered
+    else if (first == null) {
+      if (order) -1 else 1
+    }
+    // second value is null -> inequality
+    // but order is considered
+    else if (second == null) {
+      if (order) 1 else -1
+    }
+    // no null values
+    else {
+      wrappedComparator.compare(first, second)
+    }
+  }
+
+  override def compareToReference(referencedComparator: TypeComparator[T]): Int = {
+    val otherComparator = referencedComparator.asInstanceOf[NullAwareComparator[T]]
+    val otherNullReference = otherComparator.nullReference
+    // both values are null -> equality
+    if (nullReference && otherNullReference) {
+      0
+    }
+    // first value is null -> inequality
+    // but order is considered
+    else if (nullReference) {
+      if (order) 1 else -1
+    }
+    // second value is null -> inequality
+    // but order is considered
+    else if (otherNullReference) {
+      if (order) -1 else 1
+    }
+    // no null values
+    else {
+      wrappedComparator.compareToReference(otherComparator.wrappedComparator)
+    }
+  }
+
+  override def supportsNormalizedKey(): Boolean = wrappedComparator.supportsNormalizedKey()
+
+  override def equalToReference(candidate: T): Boolean = {
+    // both values are null
+    if (candidate == null && nullReference) {
+      true
+    }
+    // one value is null
+    else if (candidate == null || nullReference) {
+      false
+    }
+    // no null value
+    else {
+      wrappedComparator.equalToReference(candidate)
+    }
+  }
+
+  override def duplicate(): TypeComparator[T] = {
+    new NullAwareComparator[T](wrappedComparator.duplicate(), order)
+  }
+
+  /**
+   * This method does not support records that are null. Use [[extractKeysFromNull()]] instead.
```

Comment:
We can get the number of fields to extract by calling `getFlatComparators()` in the constructor of the `NullAwareComparator` and use this number to fill the target array with nulls if `record` is `null`. 
Then we do not need the special `extractKeysFromNull` method and avoid some overhead.